Pull request for supervisor review (#1)
AstroViking/ba-thesis: chapters/01-introduction.Rmd

## Kernel Density Estimation

> In many fields of science and in physics especially, scientists need to estimate the probability density function (PDF) from which a set of data is drawn. If the underlying mechanisms are well understood a so called parametric approach can be used. One can describe the model mathematically and fit it to the experimental data by some goodness-of-fit criterion like log-likelihood or $\chi^2$. 

Reviewer: [5] Extend this even more. E.g. mention along the lines of: 
- PDFs can be used for statistical inference.
- sometimes the process can be described analytically
- rearrange the chapter: start out with the problem that we have data, want to describe it with a function that contains parameters (which can have a physical meaning). We have a large amount of data. We do this also in Python as it grows more and more. Therefore zfit. It uses TensorFlow as the backend. Describe a little bit TF (it runs first time with algebraic expressions that are faster to evaluate afterwards), and that it is therefore not compatible directly. However, it can still be used together with other libraries (e.g. it takes numpy arrays).
- maybe also outline here that _exact_ implementations exists in TF (probability) but that these are inefficient for larger data samples

> Oftentimes however the underlying mechanisms are too complex to be fully understood analytically and the parametric methods fail. The knowledge of the system is to poor to describe the model as mathematical function.

Reviewer: ```suggestion
Oftentimes however the underlying mechanisms are too complex to be fully described analytically and the parametric methods cannot be used. The knowledge of the system is to poor to describe the model as an exact mathematical function.
```

> In the particle accelerator at CERN for instance, they record a whooping 25 gigabytes of data per second[@cern], resulting from the process of many physical interactions that occur almost simultaneously. It is not feasible to anticipate features of the distribution one observes experimentally, as it is comprised of many different distributions, which result from all the different physical interactions.

Reviewer: ```suggestion
In the particle accelerator at CERN for instance, they record a whooping 25 gigabytes of data per second[@cern], resulting from the process of many physical interactions that occur almost simultaneously. It is often not feasible to anticipate features of the distribution one observes experimentally, as it is comprised of many different distributions, which result from all the different physical interactions. However, through Monte Carlo based simulation, samples can be drawn from the theoretical distribution. These can be used in conjunction with non-parametric methods that approximate the shape of the drawn distribution.
```

To combat this, so called non-parametric methods are used, which do not need a predefined mathematical model. 

The perhaps simplest non-parametric method is the Histogramm. By summing the the data up in discrete bins the underlying parent distribution can be approximated, without needing any knowledge of underlying processes.

> However there are also many more sophisticated non-parametric methods, one in particular is Kernel Density Estimation (KDE), which can be looked at as a sort of generalized histogram [@rosenblatt1956remarks].

Reviewer: [4] change with line 17: first talk about the drawbacks, then mention the improved versions

Histograms tend to produce PDFs that are highly dependent on bin width and bin positioning, meaning the interpretation of the data changes by a lot by two arbitrary parameters. KDE circumvents this problem by replacing each data point with a so called kernel function that specifies how much it influences its neighbouring regions as well. The kernel functions themselves are centered at the data point directly, eliminating the need for arbitrary bin positioning[@duong2001]. Since KDE still depends on kernel bandwidth (instead of bin width), one might argue that this is not a major improvement. However, upon closer inspection, one finds that the underlying PDF does depend less strongly on the kernel bandwidth than histograms on bin width and it is much easier to specify rules for an approximately optimal kernel bandwidth than it is to do so for bin width[@lerner2013]. In addition, by specifying a smooth kernel function, one gets a smooth distribution as well, which is often desirable or even expected from theory.

> Due to this increased robustness, KDE is particular useful in High-Energy Physics (HEP) where it has been used for confidence level calculations for the Higgs Searches at the Large Electron Positron Collider (LEP)[@cranmer2000kernel]. However there is still room for improvement and certain more sophisticated approaches to Kernel Density Estimation have been proposed in dependence on specific areas of application[@cranmer2000kernel].

Reviewer: [4] yes true, however it is mostly used in fits for example. I would mention this as well

## Purpose of this thesis

> The purpose of this thesis is to propose four novel implementations of Univariate kernel density estimation based on TensorFlow[@martinabadiTensorFlowLargeScaleMachine2015], TensorFlow Probability[@googleTFP] and zfit[@eschle2019zfit], which incorporate insights from recent papers to decrease the computational complexity and therefore runtime. The newly proposed implementations are then compared to the state of the art of kernel density estimation in Python and shown to be competitive. By leveraging TensorFlow's graph based computation, the newly proposed methods to calculate a kernel density estimate can benefit from parallelization and efficient computation on CPU/GPU.

Reviewer: ```suggestion
The purpose of this thesis is to propose four novel implementations of univariate kernel density estimation based on TensorFlow[@martinabadiTensorFlowLargeScaleMachine2015], TensorFlow Probability[@googleTFP] and zfit[@eschle2019zfit], which incorporate insights from recent papers to decrease the computational complexity and therefore runtime. The newly proposed implementations are then compared to the state of the art of kernel density estimation in Python and shown to be competitive. By leveraging TensorFlow's graph based computation, the newly proposed methods to calculate a kernel density estimate can benefit from parallelization and efficient computation on CPU/GPU.
```

First the mathematical theory will be summarized in chapter \@ref(mathematicalTheory), before currently existing implementations of kernel density estimation (KDE) in Python are discussed in chapter \@ref(currentState). In chapter \@ref(tensorflowImplementation) the four novel KDE implementations are proposed, which are then compared against the current state of the art in chapter \@ref(comparison). The findings are then summarized in chapter \@ref(summary).

> ## kernel density estimation

Reviewer: ```suggestion
## Kernel density estimation
```

> So far only exact kernel density estimations exist for TensorFlow, however as seen in chapter \@ref(mathematicalTheory), the KDE can be approximated using multiple mathematical tricks with only a negligible decrease in accuracy, while decreasing the computational complexity substantially.

Reviewer: ```suggestion
So far only exact KDEs exist for TensorFlow, however as seen in chapter \@ref(mathematicalTheory), the KDE can be approximated using multiple mathematical tricks with only a negligible decrease in accuracy, while decreasing the computational complexity substantially.
```
it was introduced above

> 

Reviewer: [8] In line 5: extend and make less complicated, shorter sentences. It's very dense currently and somewhat confusing: it seems like it's about multiple things at once. Explain when we can use parametric and when we can't. What we can do with non-parametric.

It seems all somewhat entangled, what exactly do you want to say in which paragraph?

The information is there, but it seems to be spread out

Currently the basic principle of KDE has been implemented in various programming languages and statistical modeling tools. The standard framework used in HEP, that includes KDE is the ROOT/RooFit toolkit written in C++. However, Python plays an increasingly large role in the natural sciences due to support by corporations involved in Big Data and its superior accessibility. To elevate research in HEP, zfit, a new alternative to RooFit, was proposed. It is implemented on top of TensorFlow, one of the leading Python frameworks to handle large data and high parallelization, allowing a transparent usage of CPUs and GPUs[@eschle2019zfit].

> So far there exists no direct implementation of Kernel Density Estimation in zfit nor TensorFlow, but various implementations in Python. This implementations will be discussed in the next chapter (\@ref(currentState)) before I propose an implementation of Kernel Density Estimation in TensorFlow for the univariate case (\@ref(tensorflowImplementation)). Starting with a rather simple implementation, multiple improvements from recent papers are integrated to ameliorate its efficiency. Efficiency and accuracy of the implementation are then tested by comparing it to other implementations in pure Python(\@ref(comparison)).

Reviewer: - [5] There is a tricky thing on which voice to use, active vs passive or (best?) both. In the end, it's a personal preference, however I would make sure to be consistent with it. E.g. to say that "This implementation will be... (passive), before I propose (active)" seems somewhat irritating, at least for me
- [5] "ameliorate" is a word I have to look up, it's good to use a decent vocabulary but the majority of researchers should be able to understand it without a dict
